src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -911,6 +917,63 @@ private void GetFlushResult(ref FlushResult result)
             }
         }
 
+        internal ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            // Allocate whatever the pool gives us so we can write, this also marks the
+            // state as writing
+            AllocateWriteHeadIfNeeded(0);
+
+            if (source.Length <= _writingMemory.Length)
+            {
+                source.CopyTo(_writingMemory);
+
+                AdvanceCore(source.Length);
+            }
+            else
+            {
+                // This is the multi segment copy
+                WriteMultiSegment(source.Span);
+            }
+
+            return FlushAsync(cancellationToken);
+        }
+
+        private void WriteMultiSegment(ReadOnlySpan<byte> source)
+        {
+            Span<byte> destination = _writingMemory.Span;
+
+            while (true)
+            {
+                int writable = Math.Min(destination.Length, source.Length);
+                source.Slice(0, writable).CopyTo(destination);
+                source = source.Slice(writable);
+                AdvanceCore(writable);
+
+                if (source.Length == 0)
+                {
+                    break;
+                }
+
+                // We filled the segment
+                _writingHead.End += writable;
+                _buffered = 0;
+
+                // This is optimized to use pooled memory. That's why we pass 0 instead of
+                // source.Length
+                BufferSegment newSegment = AllocateSegment(0);
+
+                _writingHead.SetNext(newSegment);
```

Comment:
Does `.SetNext` need to be done under lock, or is it ok as writing is active and is purely in the uncommited portion?

If not then (not for this PR); `_operationState.BeginWrite()` could be made lock-free **\*** and only the portion of `AllocateWriteHeadSynchronized` that's touching the `_readHead` and `_readTail` would need the lock?

 **\*** Avoiding race with `_operationState.IsWritingActive` in AdvanceReader would be the trickier bit

Reply:
Pretty sure it’s ok. Those are good follow up items

Reply:
Yea me too; mostly a check and follow up on @pakrym's suggestion  https://github.com/dotnet/corefx/pull/35484#discussion_r259000620